File: CustomerTestTask.Data/DynamicLinqItems/OrderItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Dynamic.Core;

namespace CustomerTestTask.Data.DynamicLinqItems
{
    public class OrderItem<TEntity> where TEntity: class
    {
        private List<string> _operatorList = new List<string>{"desc", "asc"}; 

        private string _customOperator;
        private string _name;

        public OrderItem(string name, string customOperator)
        {
            Name = name;
            CustomOperator = customOperator;        
        }

        public string Name {
            get => _name; 
            set {            
                if (typeof(TEntity).GetProperty(value) == null) 
                {
                    throw new Exception($"Invalid property name {value}");
                }
                _name = value;
            }
        }

        public string CustomOperator {
            get => _customOperator; 
            set {
                if (_operatorList.Any(o => o == value) == false)
                {
                   throw new Exception($"Invalid operator {value}");
                }
                _customOperator = value; 
            }
        }
                
        public IQueryable<TEntity> AddOrder(IQueryable<TEntity> query)
        {
            string str = $"{Name} {CustomOperator}";
            return query.Expression.Type == typeof(IOrderedQueryable<TEntity>)                 
                ? ((IOrderedQueryable<TEntity>) query).ThenBy(str) 
                : query.OrderBy(str);
        }   
    }
}

File: CustomerTestTask.Client/src/app/items/api/order.item.ts
export class OrderItem {
    public static readonly OPERATOR_LIST: string[] = [
        "desc", "asc"
    ];

    private _name: string = '';
    private _customOperator: string = '';

    constructor(
        name: string, 
        customOperator: string
    ) { 
        this.name = name;
        this.customOperator = customOperator;
    }

    public get name(): string {
        return this._name;
    }

    public set name(name: string) {        
        this._name = name;
    }

    public get customOperator(): string {
        return this._customOperator;
    }

    public set customOperator(customOperator: string) {
        if (OrderItem.OPERATOR_LIST.includes(customOperator) == false) {            
            throw Error(`invalid operator - ${customOperator}`);
        }
        
        this._customOperator = customOperator;
    }
    
    public json(): string {
        return JSON.stringify({
            name: this.name,
            customOperator: this.customOperator
        });
    }

    public static fromJson(stringValue: string): OrderItem {
        const object: any = JSON.parse(stringValue);
        return new OrderItem(object['name'], object['customOperator']); 
    }
}